### PR TITLE
Feature/lw 80

### PR DESCRIPTION
**Pull request by labiraus**
LW-80, integration with the Notify lambda

### CODE REVIEW

File: src/common/data-access/NotifyRepository.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+      addressBookId,
+      alertType,
+      personalisation,
+    });
+
+    // Adapting a keep-alive agent
+    const options = {
+      hostname: isDefined(process.env.NOTIFY_HOSTNAME, 'process.env.NOTIFY_HOSTNAME'),
+      path: isDefined(process.env.NOTIFY_PATH, 'process.env.NOTIFY_PATH') ?? '/notify',
+      port: isDefined(process.env.NOTIFY_PORT, 'process.env.NOTIFY_PORT') ?? 443,
```
Comment (robshakespeare) [inline src/common/data-access/NotifyRepository.ts:55]: The `isDefined` utility function throws an error if the value is not defined.  Sorry, my mistake for not naming it well, or providing comments.  I will fix the lack of comments and naming of the isDefined function.

However, the problem here, is that `isDefined` is being used to return a boolean, and then use null(ish?) coalescing to return a default value.  But the default value will never be used, and instead an error will be thrown, which doesn't appear to be the desired behaviour.

Long story short, the code can be changed to just:
```
      path: process.env.NOTIFY_PATH || '/notify',
      port: process.env.NOTIFY_PORT || '443',
```

The `||` operator is fine here, because it will catch nulls, `undefined` values, and empty strings.

File: src/common/data-access/NotifyRepository.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
+import https = require('https');
+import isDefined from '../utils/isDefined';
+import { Incident, Severity } from '../domain';
+
+export default class NotifyRepository {
+  raiseAlert(incident: Incident) {
```
Comment (robshakespeare) [inline src/common/data-access/NotifyRepository.ts:6]: The logic in this method should not be in this service.  Any logic around how notifications should get sent should be in the Notification service.

Please can this method just `JSON.stringify()` the whole Incident object, and send that to the Notification service.  The switch statement and related formatting logic should be moved to the Notification service.

File: src/common/data-access/NotifyRepository.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    // Mapping from Incident object to whatever is required by the email/text template
+    switch (incident.severity) {
+      case Severity.Red:
+        this.send('DES_Red', 'DES_Red', {
+          incidentId: incident.id,
+          Date: incident.timestamp,
+          Scenario: 'Driving examiner',
```
Comment (robshakespeare) [inline src/common/data-access/NotifyRepository.ts:13]: The hardcoding of `Scenario: 'Driving examiner',` part seems wrong.  Is that because the GOV Notify templates are specific to driving tests?

File: src/common/data-access/NotifyRepository.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+      path: isDefined(process.env.NOTIFY_PATH, 'process.env.NOTIFY_PATH') ?? '/notify',
+      port: isDefined(process.env.NOTIFY_PORT, 'process.env.NOTIFY_PORT') ?? 443,
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+        'Content-Length': data.length,
+      },
+    };
+
+    const request = https.request(options, result => {
```
Comment (robshakespeare) [inline src/common/data-access/NotifyRepository.ts:72]: Please change the `send` function to `async`, returning a Promise, and ideally please can you change this across to using the `fetch` library to make the async code simpler to write.

File: src/common/data-access/NotifyRepository.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import { Incident } from '../domain';
+import fetch from 'node-fetch';
+
+
+export default class NotifyRepository {
+  async raiseAlert(incident: Incident): Promise<object> {
+    console.log('Sending incident: ' + JSON.stringify(incident));
```
Comment (robshakespeare) [inline src/common/data-access/NotifyRepository.ts:7]: Please could we improve this logging statement to being `'Sending incident to Notification service: '`.  I think it'll help people when they're looking at the logs to fully understand what the log message is saying.

File: src/functions/post-incident/__tests__/handler.spec.ts [hunk 6]
```diff
@@ -9,14 +9,21 @@ import { PostIncidentModel } from '../PostIncidentModel.generated';
...
     process.env.DYNAMO_DB_TABLE_NAME = 'example-table-name';
+    process.env.NOTIFY_HOSTNAME = 'example-host-name';
+    process.env.NOTIFY_PATH = '/example-path';
+    process.env.NOTIFY_PORT = '1234';
   });
 
   afterAll(() => {
     delete process.env.DYNAMO_DB_TABLE_NAME;
+    delete process.env.NOTIFY_HOSTNAME;
+    delete process.env.NOTIFY_PATH;
```
Comment (robshakespeare) [inline src/functions/post-incident/__tests__/handler.spec.ts:25]: These `NOTIFY_` env vars can probably be removed now?

File: src/functions/post-incident/__tests__/handler.spec.ts [hunk 7]
```diff
@@ -9,14 +9,21 @@ import { PostIncidentModel } from '../PostIncidentModel.generated';
 import PostIncidentResult from '../PostIncidentResult';
 import { Incident } from '../../../common/domain';
 import { createMockAPIGatewayProxyEvent } from './support/createMockAPIGatewayProxyEvent';
+import NotifyRepository from '../../../common/data-access/NotifyRepository';
 
 describe('handleAsync', () => {
   beforeAll(() => {
     process.env.DYNAMO_DB_TABLE_NAME = 'example-table-name';
+    process.env.NOTIFY_HOSTNAME = 'example-host-name';
```
Comment (robshakespeare) [inline src/functions/post-incident/__tests__/handler.spec.ts:17]: These `NOTIFY_` env vars can probably be removed now?

File: src/integration-tests/component-tests/steps/PostIncidentSteps.ts [hunk 8]
```diff
@@ -1,146 +1,149 @@
...
-  expect(dbIncident?.statusHistory[0].updatedBy).to.eql('example-AuthdUserId-PostIncidentComponentTest');
-
-  expect(dbIncident?.timestamp).to.eql(context.inputModel.timestamp);
-  expect(dbIncident?.currentLocation?.timestamp).to.eql(context.inputModel.timestamp);
-  expect(dbIncident?.locationHistory[0]?.timestamp).to.eql(context.inputModel.timestamp);
-  expect(dbIncident?.currentStatus?.timestamp).to.eql(context.inputModel.timestamp);
-  expect(dbIncident?.statusHistory[0]?.timestamp).to.eql(context.inputModel.timestamp);
-});
+import { Given, When, Then } from 'cucumber';
+import { APIGatewayProxyResult } from 'aws-lambda';
```
Comment (robshakespeare) [inline src/integration-tests/component-tests/steps/PostIncidentSteps.ts:2]: Please can you merge in the latest changes from the `develop` branch.  There have been quite a few changes in the last few days.
